src/drivers/fluid_sdl2.c

```diff
@@ -0,0 +1,250 @@
+/* FluidSynth - A Software Synthesizer
+ *
+ * Copyright (C) 2003  Peter Hanappe and others.
+ * Copyright (C) 2018  Carlo Bramini
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * as published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA
+ */
+
+#include "fluid_synth.h"
+#include "fluid_adriver.h"
+#include "fluid_settings.h"
+
+#if SDL2_SUPPORT
+
+#include "SDL.h"
+
+typedef struct
+{
+    fluid_audio_driver_t driver;
+
+    fluid_synth_t *synth;
+    fluid_audio_callback_t write_ptr;
+
+    SDL_AudioDeviceID devid;
+
+    int frame_size;
+
+} fluid_sdl2_audio_driver_t;
+
+
+static void
+SDLAudioCallback(void *data, void *stream, int len)
+{
+    fluid_sdl2_audio_driver_t *dev = (fluid_sdl2_audio_driver_t *)data;
+
+    len /= dev->frame_size;
+
+    dev->write_ptr(dev->synth, len, stream, 0, 2, stream, 1, 2);
+}
+
+void fluid_sdl2_audio_driver_settings(fluid_settings_t *settings)
+{
+    int n, nDevs;
+
+    fluid_settings_register_str(settings, "audio.sdl2.device", "default", 0);
+    fluid_settings_add_option(settings, "audio.sdl2.device", "default");
+
+    if (SDL_Init(SDL_INIT_AUDIO))
+    {
+        FLUID_LOG(FLUID_ERR, "Failed to initialize SDL2: %s", SDL_GetError());
+        return;
+    }
+
+    nDevs = SDL_GetNumAudioDevices(0);
+
+    for (n = 0; n < nDevs; n++)
+    {
+        const char *dev_name = SDL_GetAudioDeviceName(n, 0);
+
+        if (dev_name != NULL)
+        {
+            FLUID_LOG(FLUID_DBG, "Testing audio device: %s", dev_name);
+            fluid_settings_add_option(settings, "audio.sdl2.device", dev_name);
+        }
+    }
+
+    SDL_Quit();
```

Comment:
I'm not sure how to change this, because the documentation says:

"You can use this function with atexit() to ensure that it is run when your application is shutdown, but it is not wise to do this from a library or other dynamically loaded code. "